src/python/gem5/resources/client_api/jsonclient.py

```diff
@@ -54,21 +54,59 @@ def __init__(self, path: str):
         self.path = path
         self.resources = []
 
-        if Path(self.path).is_file():
-            self.resources = json.load(open(self.path))
-        elif not self._url_validator(self.path):
-            raise Exception(
-                f"Resources location '{self.path}' is not a valid path or URL."
-            )
-        else:
-            req = request.Request(self.path)
-            try:
-                response = request.urlopen(req)
-            except URLError as e:
-                raise Exception(
-                    f"Unable to open Resources location '{self.path}': {e}"
+        try:
+            # Check if it's a valid file path
+            if Path(self.path).is_file():
+                with open(self.path, encoding="utf-8") as f:
+                    self.resources = json.load(f)
+
+            # Check if it's a valid URL
+            elif not self._url_validator(self.path):
+                raise ValueError(
+                    f"'{self.path}' is not a valid file path or URL."
                 )
-            self.resources = json.loads(response.read().decode("utf-8"))
+
+            # Handle 'file://' URIs
+            elif self.path.startswith("file://"):
+                try:
+                    path = Path(self.path.replace("file://", ""))
+                    if path.is_file():
+                        with path.open("r", encoding="utf-8") as f:
+                            self.resources = json.load(f)
+                    else:
+                        raise FileNotFoundError(
+                            f"File specified in URI '{self.path}' does not exist."
+                        )
+                except ValueError:
+                    raise ValueError(
+                        f"Invalid file URI format: '{self.path}'."
+                    )
+
+            # Handle standard URLs (HTTP/HTTPS)
+            else:
+                try:
+                    req = request.Request(self.path)
+                    with request.urlopen(req) as response:
+                        self.resources = json.loads(
+                            response.read().decode("utf-8")
+                        )
+                except URLError as e:
+                    raise ConnectionError(
+                        f"Failed to access URL '{self.path}': {e}"
+                    )
+                except json.JSONDecodeError as e:
+                    raise ValueError(
+                        f"Invalid JSON content at URL '{self.path}': {e}"
+                    )
+
+        except FileNotFoundError as e:
+            print(f"File Error: {e}")
+        except ConnectionError as e:
+            print(f"URL Error: {e}")
```

Comment:
Why is this not under the `else` "handle standard URLs"? Is there any way a connection error could happen in other cases?

Also, I generally don't like having nested try/except clauses. Could you give me an explanation for why this is necessary?

In researching this, it looks like what's bothering me about this whole code block is that it is a combination of the ["LBYL" style](https://docs.python.org/3/glossary.html#term-LBYL) and the ["EAFP" style](https://docs.python.org/3/glossary.html#term-EAFP). While the code you have here may work, I think the combination of these styles makes it difficult to understand and increases the chance of mistakes. 